runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java

```diff
@@ -99,117 +103,145 @@ public MetricsContainer getMetricsContainer(String stepName) {
-      FlinkDistributionGauge gauge = flinkDistributionGaugeCache.get(flinkMetricName);
-      if (gauge == null) {
-        gauge =
-            runtimeContext
-                .getMetricGroup()
-                .gauge(flinkMetricName, new FlinkDistributionGauge(update));
-        flinkDistributionGaugeCache.put(flinkMetricName, gauge);
-      } else {
-        gauge.update(update);
-      }
+  private <T, FlinkT> void updateMetric(
+      MetricKey key,
+      Map<String, FlinkT> flinkMetricMap,
+      TriFunction<String, MetricGroup, T, FlinkT> create,
+      BiConsumer<FlinkT, T> update,
+      T value) {
+    String flinkMetricName = getFlinkMetricNameString(key);
+
+    // update flink metric
+    FlinkT metric = flinkMetricMap.get(flinkMetricName);
+    if (metric == null) {
+      metric = create.apply(flinkMetricName, runtimeContext.getMetricGroup(), value);
+      flinkMetricMap.put(flinkMetricName, metric);
     }
+    update.accept(metric, value);
   }
 
-  private void updateGauge(Iterable<MetricResult<GaugeResult>> gauges) {
-    for (MetricResult<GaugeResult> metricResult : gauges) {
-      String flinkMetricName = getFlinkMetricNameString(metricResult.getKey());
+  private void updateCounter(MetricKey metricKey, long attempted) {
+    updateMetric(
+        metricKey,
+        flinkCounterCache,
+        (name, group, value) -> group.counter(name),
+        (counter, value) -> {
+          counter.dec(counter.getCount());
+          counter.inc(value);
+        },
+        attempted);
+  }
 
-      GaugeResult update = metricResult.getAttempted();
+  private void updateDistribution(MetricKey metricKey, DistributionResult attempted) {
+    updateMetric(
+        metricKey,
+        flinkDistributionGaugeCache,
+        (name, group, value) -> group.gauge(name, new FlinkDistributionGauge(value)),
+        FlinkDistributionGauge::update,
+        attempted);
+  }
 
-      // update flink metric
-      FlinkGauge gauge = flinkGaugeCache.get(flinkMetricName);
-      if (gauge == null) {
-        gauge = runtimeContext.getMetricGroup().gauge(flinkMetricName, new FlinkGauge(update));
-        flinkGaugeCache.put(flinkMetricName, gauge);
-      } else {
-        gauge.update(update);
-      }
-    }
+  private void updateGauge(MetricKey metricKey, GaugeResult attempted) {
+    updateMetric(
+        metricKey,
+        flinkGaugeCache,
+        (name, group, value) -> group.gauge(name, new FlinkGauge(value)),
+        FlinkGauge::update,
+        attempted);
   }
 
-  @VisibleForTesting
   static String getFlinkMetricNameString(MetricKey metricKey) {
     MetricName metricName = metricKey.metricName();
-    // We use only the MetricName here, the step name is already contained
-    // in the operator name which is passed to Flink's MetricGroup to which
-    // the metric with the following name will be added.
-    return metricName.getNamespace() + METRIC_KEY_SEPARATOR + metricName.getName();
+    return String.join(
```

Comment:
Did you check how the metric names look in Flink? I think we changed this some time ago because the metric names contained duplicate strings.

Reply:
I've not. I'll do that and report back.

Reply:
Have you had a chance to check?

Reply:
Please see https://issues.apache.org/jira/browse/BEAM-6172

Reply:
So just to confirm, including the step name here will duplicate it because it is already used as the metric group name for the Flink operator which calls out to the Metrics code here.